### PR TITLE
MAINT: rename MaximumFlowResult.residual to flow

### LINKED ISSUE #14731: BUG: Incorrect residual graph in scipy.sparse.csgraph.maximum_flow
### Describe your issue.

Consider the reproducing example below. This seems to be the incorrect residual graph from the flow. 1-2 has a 5 capacity, 2-3 has a 1 capacity and 3-0 has a 1 capacity. So, while the flow_value is 1 from 1-0 (which is correctly returned), the residual for 1-2 should be 4. Any insight on this is appreciated! 

### Reproducing Code Example

```python
from scipy.sparse.csgraph import maximum_flow
from scipy.sparse import csr_matrix
matrix = [[0,10,10,0],[0,0,5,0],[0,0,0,1],[1,0,0,0]]
obj = maximum_flow(csr_matrix(matrix), 1, 0)
obj.residual.toarray()
```

### Error message

```shell
array([[ 0,  0,  0, -1],
       [ 0,  0,  1,  0],
       [ 0, -1,  0,  1],
       [ 1,  0, -1,  0]], dtype=int32)
```

### SciPy/NumPy/Python version information

1.7.1 1.21.2 sys.version_info(major=3, minor=7, micro=2, releaselevel='final', serial=0)

### PR DESCRIPTION
**Pull request by fuglede**
The attribute `MaximumFlowResult.residual` was poorly named: It in fact
represents the flow function of the maximum flow, whereas the residual
is given by the difference between the input graph and this flow
function. As such, we simply rename it from `residual` to `flow` to
avoid the risk of confusion.

A different possible fix would be to change the attribute to actually be
the residual graph, but that would change its value and therefore lead
to a slightly trickier deprecation. The flow is also useful in itself,
so for now, we simply include that, and let it be up to the user to make
the more or less trivial calculation of the residual if that's ever
needed.

This closes https://github.com/scipy/scipy/issues/14731

I think this is the first time I add a `DeprecationWarning`, so I'd appreciate
if the prospective reviewer could take a look at the form of that.

cc @czgdp1807 @pmla

### CODE REVIEW

**Review by czgdp1807 — APPROVED** [review]

File: scipy/sparse/csgraph/_flow.pyx
Comment (tupui) [inline scipy/sparse/csgraph/_flow.pyx:35]: We are missing `stacklevel` here
```suggestion
        warnings.warn(
            "The attribute `residual` has been renamed to `flow`"
            " and will be removed in SciPy 1.10.",
            DeprecationWarning, stacklevel=2
        )
```

**Review by tupui — APPROVED** [review]
Thanks @fuglede. LGTM, just a small thing about `stacklevel`. I will commit directly and if green I will merge.

**Comment by fuglede** [comment]
Great, thanks both for checking.

**Comment by j-bowhay** [comment]
@h-vetinari a future deprecation to be added to #15765

**Comment by h-vetinari** [comment]
A removal after one release (i.e. deprecation in 1.9, removal in 1.10) is a bit aggressive, no? I think we should adapt the warning to promise removal by 1.11. Could you send a follow-up PR @fuglede?

(Thanks for the ping @j-bowhay 🙃)

**Comment by tupui** [comment]
@h-vetinari Note that our policy is flexible. In this case I did not think it was necessary. Feel free to send a quick fix otherwise.

**Comment by h-vetinari** [comment]
I know that the policy if flexible, and I get that the naming was suboptimal, but just removing the old way of doing things will result in a hard break for everyone who had warning-less code in 1.8.

It's not uncommon that people don't upgrade for a while and then bump more than one version in one go. Of course we cannot accommodate that indefinitely, but two releases (~= 1 year) seems like a decent trade-off.

I'd be less outspoken if this was just a necessary behaviour-change, but for removing something entirely, we should leave more time IMO.

I'd ask @fuglede to do this, since it was their PR and I just didn't see it in time before it got merged.

**Comment by tupui** [comment]
@h-vetinari First, a maintainer approved, and for such small thing it's more important to have a common voice than debating and have such conversation. We can meet offline to talk about these of course.

Second, it took me the same time to make a PR than it certainly took you to comment. So please, next time just do the change. If you really look at it, we have very limited people contributing to SciPy. So let's help everyone by not asking to make extra PRs to fix a single char.

**Comment by tupui** [comment]
I made the PR and updated the version. Thanks again everyone. All close here 😃

**Pull request merged**
